search/mongo_search.go

```diff
@@ -359,8 +329,136 @@ func (m *MongoSearcher) createParamObjects(params []SearchParam) []bson.M {
 }
 
 func (m *MongoSearcher) createPipelineObject(query Query) []bson.M {
-	// Currently this is just a $match stage with the same query that would be used in find()
-	return []bson.M{{"$match": m.createQueryObject(query)}}
+	standardSearchParams := []SearchParam{}
+	chainedSearchParams := []SearchParam{}
+
+	// Separate out chained search parameters
+	for _, p := range query.Params() {
+		if usesChainedSearch(p) {
+			chainedSearchParams = append(chainedSearchParams, p)
+			continue
+		}
+		standardSearchParams = append(standardSearchParams, p)
+	}
+
+	// Process standard SearchParams
+	pipeline := []bson.M{{"$match": m.createQueryObjectFromParams(standardSearchParams)}}
+
+	// Process chained search parameters
+	for _, p := range chainedSearchParams {
+		pipeline = append(pipeline, m.createChainedSearchPipelineStages(p)...)
+	}
+
+	return pipeline
+}
+
+// The SearchParam argument should be either a ReferenceParam or an OrParam.
+func (m *MongoSearcher) createChainedSearchPipelineStages(searchParam SearchParam) []bson.M {
+	// This returns 2 stages in the pipeline that represent a chained query reference:
+	// 1. A $lookup for the foreign Resource being referenced
+	// 2. A $match on that foreign Resource
+	stages := make([]bson.M, 2)
+
+	// Build the $lookup. We need to get a ReferenceParam (of type ChainedQueryReference)
+	// that we can use to populate the $lookup. If it's an OR, any one of its Items
+	// should do.
+	var lookupRef *ReferenceParam
+	var ok bool
+
+	_, isOr := searchParam.(*OrParam)
+
+	if isOr {
+		lookupRef, ok = searchParam.(*OrParam).Items[0].(*ReferenceParam)
+		if !ok {
+			panic(createInternalServerError("", ""))
+		}
+	} else {
+		lookupRef = searchParam.(*ReferenceParam)
+	}
+
+	chainedRef, ok := lookupRef.Reference.(ChainedQueryReference)
+	if !ok {
+		panic(createInternalServerError("", ""))
+	}
+
+	collectionName := models.PluralizeLowerResourceName(chainedRef.Type)
+
+	stages[0] = bson.M{"$lookup": bson.M{
+		"from":         collectionName,
+		"localField":   convertSearchPathToMongoField(lookupRef.getInfo().Paths[0].Path) + ".referenceid",
```

Comment:
No, you can't have more than one `as` field in a single lookup.  I was thinking this would be multiple `$lookup` stages in a row, each with a different `localField` value (based on the path) and different `as` value (to prevent clashing).  These multiple `$lookup` stages would be followed by a single `$match` stage that matches over all of the `as` fields.  You can't run the paths in `$lookup`/`$match` pairs because that's not how `or` logic works.

Regarding overwriting the `as` field, I guess you're right since these pipelines are executed serially.  In fact, overwriting the `as` field may improve performance by reducing memory consumption.  Of course, in the multi-path scenario described above, you can't have the adjacent `$lookup`s overwriting each other's `as` fields.